### PR TITLE
MiniMax-M2: pipeline-parallel safe forward/weight-loading + comm

### PR DESCRIPTION
**Pull request by jtreal**
## Motivation

Fixes initialization/cuda-graph capture failures when using --pp-size > 1.
Prevents partial/incorrect weight loading under PP.
Aligns MiniMax-M2 model code with current communicator API.

## Modifications

Make MiniMax-M2 model forward pipeline-parallel aware: Non-first PP stages consume pp_proxy_tensors.
Non-last PP stages return PPProxyTensors and skip logits/head.
Make load_weights() robust under PP when checkpoint iterator yields global keys: Skip only parameters that are expected to be absent on this PP rank. Keep raising on unexpected missing keys (avoids silently skipping real mapping bugs).
Update LayerScatterModes.init_new(...) invocation to pass is_next_layer_sparse (required by current communicator context).

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/models/minimax_m2.py [hunk 1]
```diff
@@ -347,31 +350,22 @@ def __init__(
         self.total_num_kv_heads = config.num_key_value_heads
 
         if self.total_num_kv_heads >= tp_size:
-            # Number of KV heads is greater than TP size, so we partition
```
Comment (rogeryoungh) [inline python/sglang/srt/models/minimax_m2.py:350]: Let's keep the original comments as they are helpful for maintenance.

File: python/sglang/srt/models/minimax_m2.py [hunk 2]
```diff
@@ -515,16 +505,29 @@ def __init__(
...
+            # Newer API
+            self.layer_scatter_modes = LayerScatterModes.init_new(
+                layer_id=layer_id,
+                num_layers=config.num_hidden_layers,
+                is_layer_sparse=self.is_layer_sparse,
+                is_previous_layer_sparse=is_previous_layer_sparse,
+                is_next_layer_sparse=is_next_layer_sparse,
+            )
+        except TypeError:
+            # Older API (no is_next_layer_sparse)
```
Comment (rogeryoungh) [inline python/sglang/srt/models/minimax_m2.py:522]: The older API implementation for LayerScatterModes doesn't seem to exist in the current codebase? We can probably remove this part.

File: python/sglang/srt/models/minimax_m2.py [hunk 3]
```diff
@@ -790,32 +787,65 @@ def forward(
         positions: torch.Tensor,
         forward_batch: ForwardBatch,
         input_embeds: torch.Tensor = None,
-    ) -> torch.Tensor:
-        # _print_tensor_info(input_ids, "input_ids")
-        hidden_states = self.model(input_ids, positions, forward_batch, input_embeds)
+        pp_proxy_tensors: Optional[PPProxyTensors] = None,
+    ) -> Union[torch.Tensor, PPProxyTensors]:
+        pp_group = get_pp_group()
```
Comment (rogeryoungh) [inline python/sglang/srt/models/minimax_m2.py:792]: It's better to move `get_pp_group()` to `__init__` instead of calling it in forward.

File: python/sglang/srt/models/minimax_m2.py [hunk 4]
```diff
@@ -790,32 +787,65 @@ def forward(
...
         )
 
+    def _pp_should_skip_param(self, param_name: str) -> bool:
+        pp_group = get_pp_group()
+        if pp_group is None or pp_group.world_size <= 1:
+            return False
+
+        m = _LAYER_RE.match(param_name)
+        if m is not None:
+            layer_id = int(m.group(1))
```
Comment (rogeryoungh) [inline python/sglang/srt/models/minimax_m2.py:824]: We can use the existing `get_layer_id` helper here to stay consistent with other model implementations.

**Review by rogeryoungh — COMMENTED** [review]
Thanks for the PR! I've left a few suggestions above regarding the implementation details.

**Comment by hnyls2002** [comment]
@jtreal pipeline-parallel support for MiniMax-M2 landed on main via #19577, so this is superseded. Please reopen if I've missed something.
